### PR TITLE
fixes #2400 continuation - PG constraints deferred and change destroy_all to delete_all

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

File: db/migrate/20130908100439_delete_orphaned_records.rb [hunk 1]
```diff
@@ -16,17 +16,17 @@ def up
...
-    Token.where("host_id NOT IN (?)", Host::Base.pluck(:id)).destroy_all
-    TrendCounter.where("trend_id NOT IN (?)", Trend.pluck(:id)).destroy_all
+    KeyPair.where("compute_resource_id NOT IN (?)", ComputeResource.pluck(:id)).delete_all
+    LookupKey.where("puppetclass_id NOT IN (?)", Puppetclass.pluck(:id)).delete_all
+    LookupValue.where("lookup_key_id NOT IN (?)", LookupKey.pluck(:id)).delete_all
+    FactValue.where("fact_name_id NOT IN (?) OR host_id NOT IN (?)", FactName.pluck(:id), Host::Base.pluck(:id)).delete_all
+    TaxableTaxonomy.where("taxonomy_id NOT IN (?)", Taxonomy.pluck(:id)).delete_all
+    HostClass.where("host_id NOT IN (?) OR puppetclass_id NOT IN (?)", Host::Base.pluck(:id), Puppetclass.pluck(:id)).delete_all
+    HostgroupClass.where("hostgroup_id NOT IN (?) OR puppetclass_id NOT IN (?)", Hostgroup.pluck(:id), Puppetclass.pluck(:id)).delete_all
+    Log.where("message_id NOT IN (?) OR report_id NOT IN (?) OR source_id NOT IN (?)", Message.pluck(:id), Report.pluck(:id), Source.pluck(:id)).delete_all
```
Comment (domcleal) [inline db/migrate/20130908100439_delete_orphaned_records.rb]: If we switch to delete_all, we need to check the ordering of these as we won't have callbacks.

If we destroy logs before reports, we could end up with orphaned logs if reports get deleted as they don't have a host ID.  I haven't checked for other instances in this migration, but this one jumped out at me.

**Comment by domcleal** [comment]
ping @isratrade

**Comment by isratrade** [comment]
New commit (that I will squash after your review) with changes to order of rows in 'delete_orphaned_records' migration file.
I checked the objects that I run .delete_all and the only ones with callbacks :dependant => :destory are:

``` ruby
class LookupKey
  has_many environment_classes, :dependent => :destroy
  has_many lookup_values, :dependent => :destroy

class Report
  has_many :logs, :dependent => :destroy
```

I changed the order to to ensure the migration won't fail.  If there are orphaned keys, the migration will fail since there are already foreign key constraints on the database.

**Comment by domcleal** [comment]
Thanks @isratrade, merged as c520fc0!

**Pull request closed without merging**